File: sdk/ruby/lib/agentdyne/http.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "openssl"

module AgentDyne
  # Internal HTTP client. Not intended for direct use.
  class HttpClient # :nodoc:
    SDK_VERSION    = AgentDyne::VERSION
    DEFAULT_TIMEOUT  = 60
    NON_RETRYABLE   = [400, 401, 403, 404, 422].freeze

    def initialize(api_key:, base_url:, timeout:, max_retries:)
      @api_key     = api_key
      @base_url    = URI(base_url.chomp("/"))
      @timeout     = timeout
      @max_retries = max_retries
    end

    def get(path, params = {})
      request_with_retry(:get, path, params: params)
    end

    def post(path, body = nil)
      request_with_retry(:post, path, body: body)
    end

    def patch(path, body = nil)
      request_with_retry(:patch, path, body: body)
    end

    def delete(path)
      request_with_retry(:delete, path)
    end

    # Yields raw SSE data lines.
    def stream(path, body, &block)
      uri = build_uri(path)
      http = build_http(uri)

      req = Net::HTTP::Post.new(uri)
      apply_headers(req, stream: true)
      req.body = JSON.generate(body)

      http.request(req) do |resp|
        raise_for_status(resp) unless resp.is_a?(Net::HTTPSuccess)
        resp.read_body do |chunk|
          chunk.each_line do |line|
            line = line.strip
            next unless line.start_with?("data: ")
            data = line[6..]
            return if data == "[DONE]"
            block.call(data)
          end
        end
      end
    rescue Errno::ECONNREFUSED, Errno::ETIMEDOUT, SocketError => e
      raise NetworkError.new(e.message, cause: e)
    end

    private

    def request_with_retry(method, path, params: {}, body: nil)
      last_error = nil

      (@max_retries + 1).times do |attempt|
        if attempt > 0
          sleep(backoff_delay(attempt - 1))
        end

        begin
          result = execute_request(method, path, params: params, body: body)
          return result
        rescue RateLimitError => e
          sleep(e.retry_after_seconds) if attempt < @max_retries
          last_error = e
        rescue AgentDyneError => e
          raise if NON_RETRYABLE.include?(e.status_code)
          raise if e.status_code && e.status_code < 500
          last_error = e
        rescue NetworkError => e
          last_error = e
        end
      end

      raise last_error
    end

    def execute_request(method, path, params: {}, body: nil)
      uri  = build_uri(path, params)
      http = build_http(uri)
      req  = build_request(method, uri, body)

      resp = http.request(req)
      raise_for_status(resp)

      body_str = resp.body
      return nil if body_str.nil? || body_str.empty?
      JSON.parse(body_str)
    rescue Errno::ECONNREFUSED, Errno::ETIMEDOUT, SocketError => e
      raise NetworkError.new(e.message, cause: e)
    end

    def build_uri(path, params = {})
      uri = URI.join(@base_url.to_s + "/", path.delete_prefix("/"))
      unless params.empty?
        uri.query = URI.encode_www_form(params.compact)
      end
      uri
    end

    def build_http(uri)
      http              = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl      = uri.scheme == "https"
      http.read_timeout = @timeout
      http.open_timeout = @timeout
      http
    end

    def build_request(method, uri, body)
      klass = {
        get:    Net::HTTP::Get,
        post:   Net::HTTP::Post,
        patch:  Net::HTTP::Patch,
        delete: Net::HTTP::Delete,
      }[method]
      req = klass.new(uri)
      apply_headers(req)
      req.body = JSON.generate(body) if body
      req
    end

    def apply_headers(req, stream: false)
      req["Authorization"] = "Bearer #{@api_key}"
      req["Content-Type"]  = "application/json"
      req["Accept"]        = stream ? "text/event-stream" : "application/json"
      req["User-Agent"]    = "agentdyne-ruby/#{SDK_VERSION}"
      req["X-SDK-Language"]= "ruby"
    end

    def raise_for_status(resp)
      return if resp.is_a?(Net::HTTPSuccess)

      status   = resp.code.to_i
      raw_body = resp.body.to_s
      body     = begin; JSON.parse(raw_body); rescue JSON::ParserError; {}; end

      message = body["error"] || body["message"] || "HTTP #{status}"
      code    = body["code"]

      case status
      when 400 then raise ValidationError.new(message, fields: body["fields"], raw: body)
      when 401 then raise AuthenticationError.new(message, raw: body)
      when 403
        if code == "SUBSCRIPTION_REQUIRED"
          raise SubscriptionRequiredError.new(nil, raw: body)
        end
        raise PermissionDeniedError.new(message, raw: body)
      when 404 then raise NotFoundError.new(message, raw: body)
      when 429
        if code == "QUOTA_EXCEEDED"
          raise QuotaExceededError.new(nil, raw: body)
        end
        retry_after = (resp["Retry-After"] || "60").to_f
        raise RateLimitError.new(retry_after, raw: body)
      else
        if status >= 500
          raise InternalServerError.new(message, raw: body)
        end
        raise AgentDyneError.new(message, status_code: status, code: code, raw: body)
      end
    end

    def backoff_delay(attempt)
      base    = 0.5
      cap     = 30.0
      ceiling = [cap, base * (2**attempt)].min
      rand * ceiling
    end
  end
end

File: sdk/python/agentdyne/http.py
"""
http.py — HTTP transport for the AgentDyne Python SDK.

Sync: pure stdlib (urllib) — zero dependencies.
Async: httpx (optional, installed with `pip install agentdyne[async]`).
"""

from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, AsyncIterator, Dict, Generator, Iterator, Optional

from .errors import (
    AgentDyneError,
    AuthenticationError,
    NotFoundError,
    QuotaExceededError,
    RateLimitError,
    ServerError,
    ValidationError,
)

# Try to import httpx for async support
try:
    import httpx as _httpx
    _HTTPX_AVAILABLE = True
except ImportError:
    _httpx = None        # type: ignore
    _HTTPX_AVAILABLE = False

_DEFAULT_HEADERS = {
    "User-Agent":   "agentdyne-python/2.0.0",
    "Content-Type": "application/json",
    "Accept":       "application/json",
}

_RETRY_STATUS = {429, 500, 502, 503, 504}


def _raise_for_status(status: int, body: bytes, url: str) -> None:
    try:
        data = json.loads(body) if body else {}
        message = data.get("error") or data.get("message") or str(body[:200])
    except Exception:
        message = str(body[:200])

    if status == 401:
        raise AuthenticationError(message)
    if status == 404:
        raise NotFoundError(message)
    if status == 422:
        raise ValidationError(message)
    if status == 429:
        raise RateLimitError(message, retry_after=None)
    if status == 402:
        raise QuotaExceededError(message)
    if status >= 500:
        raise ServerError(message, status_code=status)
    if status >= 400:
        raise AgentDyneError(message, status_code=status)


class HttpClient:
    """Synchronous HTTP client — pure stdlib, zero deps."""

    def __init__(self, *, api_key: str, base_url: str, timeout: float, max_retries: int) -> None:
        self._api_key    = api_key
        self._base_url   = base_url.rstrip("/")
        self._timeout    = timeout
        self._max_retries = max_retries

    def _headers(self) -> Dict[str, str]:
        return {**_DEFAULT_HEADERS, "X-API-Key": self._api_key}

    def _url(self, path: str, params: Optional[Dict[str, Any]] = None) -> str:
        url = f"{self._base_url}{path}"
        if params:
            url += "?" + urllib.parse.urlencode({k: v for k, v in params.items() if v is not None})
        return url

    def _request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Dict[str, Any]] = None,
        body:   Optional[Dict[str, Any]] = None,
    ) -> Any:
        url     = self._url(path, params)
        data    = json.dumps(body).encode("utf-8") if body is not None else None
        headers = self._headers()
        if data:
            headers["Content-Length"] = str(len(data))

        for attempt in range(self._max_retries + 1):
            try:
                req = urllib.request.Request(url, data=data, headers=headers, method=method)
                with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                    resp_body = resp.read()
                    return json.loads(resp_body) if resp_body else {}
            except urllib.error.HTTPError as exc:
                resp_body = exc.read()
                if exc.code in _RETRY_STATUS and attempt < self._max_retries:
                    time.sleep(2 ** attempt)
                    continue
                _raise_for_status(exc.code, resp_body, url)
                raise  # unreachable — _raise_for_status always raises
            except urllib.error.URLError as exc:
                if attempt < self._max_retries:
                    time.sleep(2 ** attempt)
                    continue
                raise AgentDyneError(f"Request failed: {exc.reason}") from exc

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Any:
        return self._request("GET", path, params=params)

    def post(self, path: str, body: Dict[str, Any]) -> Any:
        return self._request("POST", path, body=body)

    def patch(self, path: str, body: Optional[Dict[str, Any]] = None) -> Any:
        return self._request("PATCH", path, body=body or {})

    def delete(self, path: str) -> Any:
        return self._request("DELETE", path)

    def stream(self, path: str, body: Dict[str, Any]) -> Iterator[str]:
        """Yield raw SSE data lines from a streaming endpoint."""
        url  = self._url(path)
        data = json.dumps(body).encode("utf-8")
        headers = {**self._headers(), "Accept": "text/event-stream", "Content-Length": str(len(data))}
        req = urllib.request.Request(url, data=data, headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                for raw_line in resp:
                    line = raw_line.decode("utf-8").strip()
                    if line.startswith("data: "):
                        text = line[6:]
                        if text == "[DONE]":
                            return
                        yield text
        except urllib.error.HTTPError as exc:
            resp_body = exc.read()
            _raise_for_status(exc.code, resp_body, url)


class AsyncHttpClient:
    """Async HTTP client — requires httpx."""

    def __init__(self, *, api_key: str, base_url: str, timeout: float, max_retries: int) -> None:
        if not _HTTPX_AVAILABLE:
            raise ImportError("AsyncAgentDyne requires httpx. Install with: pip install agentdyne[async]")
        self._api_key     = api_key
        self._base_url    = base_url.rstrip("/")
        self._timeout     = timeout
        self._max_retries = max_retries
        self._client: Any = _httpx.AsyncClient(timeout=timeout)

    def _headers(self) -> Dict[str, str]:
        return {**_DEFAULT_HEADERS, "X-API-Key": self._api_key}

    def _url(self, path: str, params: Optional[Dict[str, Any]] = None) -> str:
        url = f"{self._base_url}{path}"
        if params:
            url += "?" + urllib.parse.urlencode({k: v for k, v in params.items() if v is not None})
        return url

    async def _request(
        self,
        method: str,
        path: str,
        *,
        params: Optional[Dict[str, Any]] = None,
        body:   Optional[Dict[str, Any]] = None,
    ) -> Any:
        url = self._url(path, params)
        for attempt in range(self._max_retries + 1):
            try:
                resp = await self._client.request(
                    method, url,
                    headers=self._headers(),
                    json=body,
                )
                if resp.status_code in _RETRY_STATUS and attempt < self._max_retries:
                    await _httpx.sleep(2 ** attempt)  # type: ignore
                    continue
                _raise_for_status(resp.status_code, resp.content, url)
                return resp.json() if resp.content else {}
            except _httpx.RequestError as exc:
                if attempt < self._max_retries:
                    continue
                raise AgentDyneError(f"Request failed: {exc}") from exc

    async def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Any:
        return await self._request("GET", path, params=params)

    async def post(self, path: str, body: Dict[str, Any]) -> Any:
        return await self._request("POST", path, body=body)

    async def patch(self, path: str, body: Optional[Dict[str, Any]] = None) -> Any:
        return await self._request("PATCH", path, body=body or {})

    async def delete(self, path: str) -> Any:
        return await self._request("DELETE", path)

    async def stream(self, path: str, body: Dict[str, Any]) -> AsyncIterator[str]:
        url = self._url(path)
        async with self._client.stream("POST", url, headers={**self._headers(), "Accept": "text/event-stream"}, json=body) as resp:
            async for raw_line in resp.aiter_lines():
                line = raw_line.strip()
                if line.startswith("data: "):
                    text = line[6:]
                    if text == "[DONE]":
                        return
                    yield text

    async def aclose(self) -> None:
        await self._client.aclose()
